examples/applications/plot_cyclical_feature_engineering.py

```diff
@@ -0,0 +1,693 @@
+# equivalent information in a non-monotonic way, and more importantly without
+# any jump between the first and the last value of the periodic range.
+from sklearn.preprocessing import FunctionTransformer
+
+
+def sin_transformer(period):
+    return FunctionTransformer(lambda x: np.sin(x / period * 2 * np.pi))
+
+
+def cos_transformer(period):
+    return FunctionTransformer(lambda x: np.cos(x / period * 2 * np.pi))
+
+
+# %%
+#
+# Let us visualize the effect of this feature expansion on some synthetic hour
+# data with a bit of extrapolation beyond hour=23:
+import pandas as pd
+
+hour_df = pd.DataFrame(
+    np.linspace(0, 26, 1000).reshape(-1, 1),
+    columns=["hour"],
+)
+hour_df["hour_sin"] = sin_transformer(24).fit_transform(hour_df)["hour"]
+hour_df["hour_cos"] = cos_transformer(24).fit_transform(hour_df)["hour"]
+hour_df.plot(x="hour")
+_ = plt.title("Trigonometric encoding for the 'hour' feature")
+
+# %%
+#
+# We can now build a feature extraction pipeline using this strategy:
+cyclic_cossin_transformer = ColumnTransformer(
+    transformers=[
+        ("categorical", one_hot_encoder, categorical_columns),
+        ("month_sin", sin_transformer(12), ["month"]),
+        ("month_cos", cos_transformer(12), ["month"]),
+        ("weekday_sin", sin_transformer(7), ["weekday"]),
+        ("weekday_cos", cos_transformer(7), ["weekday"]),
+        ("hour_sin", sin_transformer(24), ["hour"]),
+        ("hour_cos", cos_transformer(24), ["hour"]),
+    ],
+    remainder="passthrough",
+)
+cyclic_cossin_linear_pipeline = make_pipeline(
+    cyclic_cossin_transformer,
+    RidgeCV(alphas=alphas),
+)
+evaluate(cyclic_cossin_linear_pipeline, X, y, cv=ts_cv)
+
+
+# %%
+#
+# The performance of our linear regression model with this simple feature
+# engineering is a bit better than using the original ordinal time features but
+# worse than using the one-hot encoded time features. We will further analyze
+# possible reasons for this disappointing outcome at the end of this notebook.
+#
+# Periodic spline features
+# ------------------------
+#
+# We can try an alternative encoding of the periodic time-related features
+# using spline transformations with a large-enough number of knots, and as a
+# result a larger number of expanded features:
+from sklearn.preprocessing import SplineTransformer
+
+
+def periodic_spline_transformer(period, n_knots=None, degree=3):
+    if n_knots is None:
+        n_knots = period
+    return SplineTransformer(
+        degree=degree,
+        n_knots=n_knots,
+        knots=np.linspace(0, period, n_knots).reshape(n_knots, 1),
+        extrapolation="periodic",
+    )
+
+
+# %%
+#
+# Again, let us visualize the effect of this feature expansion on some
+# synthetic hour data with a bit of extrapolation beyond hour=23:
+hour_df = pd.DataFrame(
+    np.linspace(0, 26, 1000).reshape(-1, 1),
+    columns=["hour"],
+)
+splines = periodic_spline_transformer(24, n_knots=12).fit_transform(hour_df)
```

Comment:
I see what I can do. It makes me think that maybe the `SplineTransformer` should allow for `n_splines` and a `period` argument...

But that could make the parameters docstring very complex to understand.

Reply:
When writing the `SplineTransformer`, I thought the number of knots is more intuitive than the number of splines/dof. But I documented the numbers very clearly (and it is accesible via `n_features_out_)`. I did not, however, think of periodic splines or a period. That came only a little later with @mlondschien.

A `period` argument, however, would make sense in my opinion.

Reply:
I was only suggesting to use `periodic_spline_transformer(24, n_knots=13)` in this example, to get 12 splines instead of 11. I agree the numbers are well documented in `SplineTransformer`.

Reply:
Nice writeup! Different to what you are writing above @ogrisel, the number of knots you are choosing are not natural. They are arbitrary. I would vary the period but keep the number of knots fixed for month / weekday / hour (e.g. 5?). If you use `period + 1` knots, resulting in `period` splines, the resulting splines are equivalent as using one-hot-encoded features (assuming integer value features). This is why the performance of splines is so similar to one-hot-encoded features. To benefit from the additional "smoothness" from splines, you will need to reduce the number of splines. Note that you could use non-evenly spaced knots, e.g. via quantiles.

Reply:
If you want to display the strengths of periodic splines, I would suggest to include interactions between periodic transformations of the time variables. For e.g. 4 knots this could be manageable, whereas this would explode for one-hot encoded features.

Reply:
I updated the example to control for the number of splines and made the number of knots a technical detail.